extern/sector-storage/sched.go

```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
+		active:    &activeResources{},
+		enabled:   true,
+
+		closingMgr: make(chan struct{}),
+		closedMgr:  make(chan struct{}),
+	}
+
+	wid := WorkerID(sessID)
+
+	sh.workersLk.Lock()
+	_, exist := sh.workers[wid]
+	if exist {
+		// this is ok, we're already handling this worker in a different goroutine
+		return nil
+	}
+
+	sh.workers[wid] = worker
+	sh.workersLk.Unlock()
+
+	go func() {
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
 
 		defer close(worker.closedMgr)
 
 		scheduledWindows := make(chan *schedWindow, SchedWindows)
 		taskDone := make(chan struct{}, 1)
 		windowsRequested := 0
 
-		ctx, cancel := context.WithCancel(context.TODO())
-		defer cancel()
+		disable := func(ctx context.Context) error {
+			done := make(chan struct{})
 
-		workerClosing, err := worker.w.Closing(ctx)
-		if err != nil {
-			return
+			// request cleanup in the main scheduler goroutine
+			select {
+			case sh.workerDisable <- workerDisableReq{
+				activeWindows: worker.activeWindows,
+				wid:           wid,
+				done: func() {
+					close(done)
+				},
+			}:
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-sh.closing:
+				return nil
+			}
+
+			// wait for cleanup to complete
+			select {
+			case <-done:
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-sh.closing:
+				return nil
+			}
+
+			worker.activeWindows = worker.activeWindows[:0]
+			windowsRequested = 0
+			return nil
 		}
 
 		defer func() {
-			log.Warnw("Worker closing", "workerid", wid)
+			log.Warnw("Worker closing", "workerid", sessID)
 
-			// TODO: close / return all queued tasks
+			if err := disable(ctx); err != nil {
+				log.Warnw("failed to disable worker", "worker", wid, "error", err)
+			}
+
+			sh.workersLk.Lock()
+			delete(sh.workers, wid)
+			sh.workersLk.Unlock()
 		}()
 
+		heartbeatTimer := time.NewTicker(stores.HeartbeatInterval)
+		defer heartbeatTimer.Stop()
+
 		for {
-			// ask for more windows if we need them
-			for ; windowsRequested < SchedWindows; windowsRequested++ {
+			sh.workersLk.Lock()
+			enabled := worker.enabled
+			sh.workersLk.Unlock()
+
+			// ask for more windows if we need them (non-blocking)
+			for ; enabled && windowsRequested < SchedWindows; windowsRequested++ {
 				select {
 				case sh.windowRequests <- &schedWindowRequest{
 					worker: wid,
 					done:   scheduledWindows,
 				}:
 				case <-sh.closing:
 					return
-				case <-workerClosing:
-					return
 				case <-worker.closingMgr:
 					return
 				}
 			}
 
-			select {
-			case w := <-scheduledWindows:
-				worker.wndLk.Lock()
-				worker.activeWindows = append(worker.activeWindows, w)
-				worker.wndLk.Unlock()
-			case <-taskDone:
-				log.Debugw("task done", "workerid", wid)
-			case <-sh.closing:
-				return
-			case <-workerClosing:
-				return
-			case <-worker.closingMgr:
-				return
+			// wait for more windows to come in, or for tasks to get finished (blocking)
+			for {
+
+				// first ping the worker and check session
+				{
```

Comment:
Lets break this subscope into a separate function, its really hard to tell whats being done here, and what the potential outcomes are (and why). I see several returns, a continue, and a few routes that just flow through. We should probably have a function that returns some value to signal whether to loopcontinue, return, or keep going